### PR TITLE
Clean up type declarations and add ability to store subfields for blocks, etc

### PR DESCRIPTION
**Pull request by evanphx**
This cleans up the type names exported, removing `*` for pointers and transforming `[]` into `list of` and `map[]` into `map of X to Y`
This also adds the ability for individual fields to have fields themself, ie subfields. This is most commonly used for configuration blocks. This allows infinite nesting, but users should be careful to use this new power responsibly.

### CODE REVIEW

File: docs/docs.go [hunk 1]
```diff
@@ -7,33 +7,77 @@ import (
...
+	Category bool
+
+	// SubFields is defined when this field is a category. It is the fields
+	// in that category.
+	SubFields []*FieldDocs
+
+	discoveredFields map[string]*FieldDocs
 }
 
+// Documentation allows a plugin to document it's many wonderful features.
```
Comment (briancain) [inline docs/docs.go:80]:
```suggestion
// Documentation allows a plugin to document its many wonderful features.
```

**Review by briancain — APPROVED** [review]
Awesome! The auto-conversion of var types will read so much better 👏🏻

File: docs/docs.go [hunk 2]
```diff
@@ -63,12 +110,19 @@ func New(opts ...Option) (*Documentation, error) {
 	return &d, nil
 }
 
+// FromConfig populates the Documentation value by reading the struct
+// members on the value. This is typically passed the value that is returned
+// by the plugins Config function.
```
Comment (krantzinator) [inline docs/docs.go:115]:
```suggestion
// by the plugin's Config function.
```

File: docs/docs.go [hunk 3]
```diff
@@ -63,12 +110,19 @@ func New(opts ...Option) (*Documentation, error) {
...
 func FromConfig(v interface{}) Option {
 	return func(d *Documentation) error {
 		return fromConfig(v, d.fields)
 	}
 }
 
+// RequestFromStruct populates the Documentation's request information
+// by reading the struct members on the value. Request information is
+// configuration defined by a Config Sourcer to be used as authentication
```
Comment (krantzinator) [inline docs/docs.go:124]:
```suggestion
// configuration defined by a ConfigSourcer to be used as authentication
```

File: docs/docs.go [hunk 4]
```diff
@@ -184,33 +273,112 @@ func (d *Documentation) AddMapper(input, output, description string) {
...
+		fields: make(map[string]*FieldDocs),
+	}
+
+	f(sf)
+
+	return sf
+}
+
+// SetField sets various documentation for the given field. If the field is already
+// known, the documentation is mearly updated. If the field is missing, it is created.
```
Comment (krantzinator) [inline docs/docs.go:360]:
```suggestion
// known, the documentation is merely updated. If the field is missing, it is created.
```

File: docs/docs.go [hunk 5]
```diff
@@ -184,33 +273,112 @@ func (d *Documentation) AddMapper(input, output, description string) {
...
 	}
 
+	applyOpts(field, opts)
+
 	return nil
 }
 
-func (d *Documentation) SetTemplateField(name, synposis string, opts ...DocOption) error {
+// SetTemplateField sets various documentation for the given template field.
+// If the field is already known, the documentation is mearly updated.
```
Comment (krantzinator) [inline docs/docs.go:379]:
```suggestion
// If the field is already known, the documentation is merely updated.
```

File: docs/docs.go [hunk 6]
```diff
@@ -222,21 +390,15 @@ func (d *Documentation) SetTemplateField(name, synposis string, opts ...DocOptio
...
-		}
-	}
+	applyOpts(field, opts)
 
 	return nil
 }
 
-func (d *Documentation) SetRequestField(name, synposis string, opts ...DocOption) error {
+// SetRequestField sets various documentation for the given request field.
+// If the field is already known, the documentation is mearly updated.
```
Comment (krantzinator) [inline docs/docs.go:399]:
```suggestion
// If the field is already known, the documentation is merely updated.
```

File: docs/docs.go [hunk 7]
```diff
@@ -305,6 +463,7 @@ func (d *Documentation) Fields() []*FieldDocs {
 	return fields
 }
 
+// Fields returns the formatted FieldDocs values for the template fields
```
Comment (krantzinator) [inline docs/docs.go:466]:
```suggestion
// TemplateFields returns the formatted FieldDocs values for the template fields
```

File: docs/docs.go [hunk 8]
```diff
@@ -320,6 +479,7 @@ func (d *Documentation) TemplateFields() []*FieldDocs {
 	return fields
 }
 
+// Fields returns the formatted FieldDocs values for the request fields
```
Comment (krantzinator) [inline docs/docs.go:482]:
```suggestion
// RequestFields returns the formatted FieldDocs values for the request fields
```

**Review by krantzinator — APPROVED** [review]
I have now read the word `field` so many times that it looks weird to my brain.
Yay docs! 🎉

**Pull request merged**